### ISSUE TITLE
sumo serilog appender failing to substitute template string variables

### ISSUE THREAD

**Issue by eddit0r** [SumoLogic/sumologic-net-appenders#81]
A sumo customer has an issue implementing the sumo serilog appender.

The sumo appender is not substituting variables as expected.

The standard serilog console logger is performing the substitutions as expected.

The customer suspects a template issue:
![image](https://user-images.githubusercontent.com/4442512/53927955-91866e80-40dc-11e9-9bfc-aa5851ade0dc.png)

example of failed substitution.
`_messageLogger.Error("{name} - Failed to publish a message, RequestId:{correlationId}, HttpResponseMessage: {httpResponse}", "PublishCandidate", requestId, errorMessage);`

It is hooked up to the sumo driver in the following way :

```
logger = new LoggerConfiguration() 
.WriteTo.RollingFile(log["LogFilePath"].ToString()) 
.MinimumLevel.Information() 
.MinimumLevel.Override("Microsoft", LogEventLevel.Warning) 
.Enrich.FromLogContext() 
.WriteTo.Console() 
.WriteTo.BufferedSumoLogic( 
new Uri(sumoConfig["MessageLogContainerUrl"]), 
sourceName: sumoConfig["MessageLogSourceName"], 
sourceCategory: sumoConfig["MessageLogSourceCategory"], 
formatter: new CompactJsonFormatter()) 
.CreateLogger();
```

**Comment by bin3377** [comment]
@wdolek any comment on this?

**Comment by eddit0r** [comment]
![image](https://user-images.githubusercontent.com/4442512/54256256-81c0cb80-45af-11e9-89fe-2288d1639162.png)
Customer reports sumo serilog driver throwing exception under load, see stack trace attached.

**Comment by wdolek** [comment]
@eddit0r Sorry for such delay in response, completely missed notification.

From how you described the issue, I'm not really sure what could be the cause - template seems to be OK, we are using logging similar way without any problem. I can try to debug, but no promise. (Yet I'm puzzled that it works as expected with just console sink)

Regarding exception thrown and its stack trace, I would assume that this is between SumoLogic server and client - as you can see, common SumoLogic code is used to send events. What does _under load_ mean in your case? Would it be possible to try with different logging framework (NLog?) under same load?

**Comment by wdolek** [comment]
@eddit0r I realized I might misunderstand your issue regarding replacing. I think what you see is intentional.

Let's say I log message with this template:
`Request to {RequestUri}, Status: {StatusCode}, Duration: {Duration} ms`

This is what I get in SumoLogic:
```
{
  "Timestamp": "2019-01-01T00:00:00.0000000Z",
  "Message": "Request to {RequestUri}, Status: {StatusCode}, Duration: {Duration} ms",
  "Level": "Information",
  "RequestUri": "https://localhost",
  "StatusCode": "200",
  "Duration": 1.000000000000001,
  "SourceContext": "MyService",
  "HttpMethod": "GET",
  "Uri": "https://localhost",
  "ActionId": "00000000-0000-0000-0000-000000000000",
  "ActionName": "MyEndpoint",
  "RequestId": "0000000000000:00000000",
  "RequestPath": "/v1/dummy",
  "CorrelationId": "00000000-0000-0000-0000-000000000000",
  "ConnectionId": "0000000000000",
  "Scope": [
    "HTTP GET https://localhost"
  ]
}
```

... so yes, `Message` does not contain replaced values - but it's actually OK - you can use it as key for lookup. Values are then set as properties (matching placeholder names) of JSON representing event. You can use these for search, creating graphs, ... 

Did this help you? Are you expecting message to be replaced? @bin3377 is this how it is supposed to work?

**Comment by bin3377** [comment]
For me, keeping value of `"Message"` field with the same result as console logger (e.g., with replaced value) sounds more reasonable. People can actually only extract `"Message"` for the essential information they needed. thought?

**Comment by wdolek** [comment]
Will try other Serilog formatters and let you know. There's nothing related to message formatting in _this_ library, but it's worth checking!

**Comment by eddit0r** [comment]
@wdolek I think the customers expectation is that the substitution would function as the console loggers did. Unfortunately I am relaying this detail secondhand. 

![image](https://user-images.githubusercontent.com/4442512/55202512-d012d300-521b-11e9-91bd-f0d051e0cb2f.png)

**Comment by wdolek** [comment]
@eddit0r, yup I see. I verified how NLog or Log4net works and indeed there's some issue for Serilog (and funnily enough I thought it's a feature!)... I will look into it and let you know.

**Comment by wdolek** [comment]
@eddit0r, so after all it's not issue of library - it's by design of Serilog `JsonFormatter` and `CompactJsonFormatter` (which you are using). In order to get placeholders replaced, please use `RenderedCompactJsonFormatter` or implement own custom formatter (more about formatting here: [Serilog/Formatting Output](https://github.com/serilog/serilog/wiki/Formatting-Output)).

Let us know whether this helped.
